botorch_community/models/gp_regression_multisource.py

```diff
@@ -0,0 +1,401 @@
+#!/usr/bin/env python3
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+r"""
+Multi-Source Gaussian Process Regression models based on GPyTorch models.
+
+References:
+
+.. [Ca2021ms]
+    Candelieri, A., & Archetti, F. (2021).
+    Sparsifying to optimize over multiple information sources:
+    an augmented Gaussian process based algorithm.
+    Structural and Multidisciplinary Optimization, 64, 239-255.
+
+Contributor: andreaponti5
+"""
+
+from __future__ import annotations
+
+from typing import Optional
+
+import torch
+
+from botorch import fit_gpytorch_mll
+from botorch.exceptions import InputDataError
+from botorch.models import FixedNoiseGP, SingleTaskGP
+from botorch.models.transforms.input import InputTransform
+from botorch.models.transforms.outcome import OutcomeTransform
+from botorch.utils import draw_sobol_samples
+from gpytorch import ExactMarginalLogLikelihood, Module
+from gpytorch.likelihoods.likelihood import Likelihood
+from gpytorch.means.mean import Mean
+from gpytorch.models import ExactGP
+from torch import Tensor
+
+
+def get_random_x_for_agp(
+    n: int,
+    bounds: Tensor,
+    q: int,
+    seed: Optional[int] = None,
+):
+    r"""Draw qMC samples from the box defined by bounds.
+    The function assures that at least one point belong to
+    the source 0 (the ground truth).
+
+    Args:
+        n: The number of samples.
+        bounds: A `2 x d` dimensional tensor specifying box constraints on a
+            `d`-dimensional space, where bounds[0, :] and bounds[1, :] correspond
+            to lower and upper bounds, respectively. The last dimension represents
+            number of sources.
+        q: The size of each q-batch.
+        seed: The seed used for initializing Owen scrambling. If None (default),
+            use a random seed.
+
+    Returns:
+        A `n x q x d`-dim tensor of qMC samples from the box
+        defined by bounds.
+    """
+    if n < 1:
+        raise InputDataError(
+            f"The number of points n should be greater than 0 (given n={n})."
+        )
+    train_x = draw_sobol_samples(bounds=bounds, n=n, q=q, seed=seed).squeeze(1)
+    train_x[..., -1] = torch.round(train_x[..., -1], decimals=0)
+    if 0 not in train_x[..., -1]:
+        true_idxs = torch.randint(0, n, [max(1, int(n * 0.2))])
+        train_x[true_idxs, -1] = 0
+    return train_x
+
+
+class SingleTaskAugmentedGP(SingleTaskGP):
+    r"""A single-task multi-source GP model.
+
+    The Augmented Gaussian Process is described in [Ca2021ms]_.
+    The basic idea is to use GP sparsification for selecting a subset of
+    the function evaluations, among those performed so far over all the
+    different sources, as inducing locations to generate the AGP approximating f(x).
+    The GP sparsification proposed is an insertion method: the set of inducing
```

Comment:
The term "inducing locations" is a bit confusing here I feel, since this is typically used for inducing points in approximate GP models. But IIUC here there aren't any approximate GP models, it's just an adaptive data selection scheme to select which inputs to use in the model (in particular, there is no interpolation or the like going on).

Reply:
Yes, it is based on a GP sparsification technique. We changed the description. I hope it's more clear now.

Reply:
thanks